[Serve] Add multiplex optimization in router

sihanwang41 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

- Add `RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S` flag to wait until the model matching.
- Choose least number of models from all replicas when there is no match.
- reduce the model id information publish interval

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

edoakes reviewed (commented):
@sihanwang41 I think if we're going to use the timeout rather than the blacklist approach for spilling back to the full replica set, I'd prefer that we entirely remove the blacklist (this is the only reason I added it).

Btw, if we don't see _any_ models with a given model_id, should we immediately spill back? (Only read the PR description and scanned the code so far, not sure if it's handled already)

sihanwang41 commented:
> @sihanwang41 I think if we're going to use the timeout rather than the blacklist approach for spilling back to the full replica set, I'd prefer that we entirely remove the blacklist (this is the only reason I added it).
> 
> Btw, if we don't see _any_ models with a given model_id, should we immediately spill back? (Only read the PR description and scanned the code so far, not sure if it's handled already)

SG, I just removed the blacklist. 
If I understand you correctly, right now, we at least wait for a time waiting for model id matching. If no matching, we will choose replicas from which have least number models.

sihanwang41 commented:
@edoakes ping for another review ^^ thx!

edoakes commented on python/ray/serve/_private/router.py:
```diff
@@ -387,40 +389,74 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
     def _get_candidate_replica_ids(
         self,
-        blacklist_replica_ids: Set[str],
-        request_metadata: Optional[RequestMetadata] = None,
     ) -> Set[str]:
-        """Get candidates from the current replica set excluding the blacklist.
+        """Get candidates from the current replica set."""
+
+        return self._replica_id_set
```
remove this method and just directly access the set

edoakes commented on python/ray/serve/_private/router.py:
```diff
@@ -387,40 +389,74 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
+        By default, we will only choose from replicas that have the requested
+        multiplexed model id, if not matched, the function will return an empty set.
+
+        If fallback_to_all_replicas is True, we will choose from all replicas,
```
this is outdated

edoakes commented on python/ray/serve/_private/constants.py:
```diff
@@ -220,3 +220,13 @@
+# Serve multiplexed matching timeout.
+# This is the timeout for the matching process of multiplexed requests. To avoid
+# thundering herd problem, the timeout value will be randomed between this value
+# and this value * 1.5. The unit is second.
```
in the code it's this * 2 ...

sihanwang41 replied:
oops. updated.

edoakes commented on python/ray/serve/_private/router.py:
```diff
@@ -385,42 +387,69 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
+                candidates = self._multiplexed_model_id_to_replica_ids[model_id]
+                if len(candidates) > 0:
+                    return candidates
+            return candidates
```
I think my previous comment wasn't addressed. As written this means that we will always wait the `RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S` on the first request when there are no replicas that have a given model. Is there any way to avoid this fixed static cost?

sihanwang41 replied:
Updated the logic. Currently "first" request will fallback immediately, and all other requests for same model id will wait for matching till  RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S.

sihanwang41 commented on python/ray/serve/handle.py:
```diff
@@ -181,6 +181,10 @@ def _options(
             stream=stream,
         )
+
+        if self._router is None:
```
eager to create router to avoid race condition that concurrent multiplexed request trigger routers creation multiple times.

edoakes replied:
pls remove from this PR as we discussed

edoakes commented on python/ray/serve/_private/constants.py:
```diff
@@ -154,7 +154,7 @@
 
 # Minimum duration to wait until broadcasting model IDs.
-PUSH_MULTIPLEXED_MODEL_IDS_INTERVAL_S = 1.0
+PUSH_MULTIPLEXED_MODEL_IDS_INTERVAL_S = 0.1
```
btw, this is going to cause a lot of additional traffic in steady state. is there a simple way to push the model_ids only on update and avoid broadcasting so frequently all the time?

sihanwang41 replied:
it won't increase unnecessary traffics. we have a guard https://github.com/ray-project/ray/blob/master/python/ray/serve/multiplex.py#L138.
This is to push the information as early as possible.
